File: 2025/day08/zig/solution.zig
const std = @import("std");

const Point = struct {
    x: i32,
    y: i32,
    z: i32,
};

const Pair = struct {
    dist_sq: i64,
    i: usize,
    j: usize,
};

fn lessThanPair(_: void, a: Pair, b: Pair) bool {
    return a.dist_sq < b.dist_sq;
}

const UnionFind = struct {
    parent: []usize,
    rank: []usize,
    size: []usize,
    allocator: std.mem.Allocator,

    fn init(allocator: std.mem.Allocator, n: usize) !UnionFind {
        const parent = try allocator.alloc(usize, n);
        const rank = try allocator.alloc(usize, n);
        const size = try allocator.alloc(usize, n);

        for (0..n) |i| {
            parent[i] = i;
            rank[i] = 0;
            size[i] = 1;
        }

        return UnionFind{
            .parent = parent,
            .rank = rank,
            .size = size,
            .allocator = allocator,
        };
    }

    fn deinit(self: *UnionFind) void {
        self.allocator.free(self.parent);
        self.allocator.free(self.rank);
        self.allocator.free(self.size);
    }

    fn find(self: *UnionFind, x: usize) usize {
        if (self.parent[x] != x) {
            self.parent[x] = self.find(self.parent[x]); // Path compression
        }
        return self.parent[x];
    }

    fn unionSets(self: *UnionFind, x: usize, y: usize) bool {
        var px = self.find(x);
        var py = self.find(y);

        if (px == py) {
            return false; // Already in same set
        }

        // Union by rank
        if (self.rank[px] < self.rank[py]) {
            const temp = px;
            px = py;
            py = temp;
        }

        self.parent[py] = px;
        self.size[px] += self.size[py];

        if (self.rank[px] == self.rank[py]) {
            self.rank[px] += 1;
        }

        return true;
    }

    fn getComponentSizes(self: *UnionFind, allocator: std.mem.Allocator) ![]usize {
        // Count components first
        var count: usize = 0;
        for (0..self.parent.len) |i| {
            if (self.parent[i] == i) {
                count += 1;
            }
        }

        const sizes = try allocator.alloc(usize, count);
        var idx: usize = 0;
        for (0..self.parent.len) |i| {
            if (self.parent[i] == i) {
                sizes[idx] = self.size[i];
                idx += 1;
            }
        }

        return sizes;
    }
};

fn euclideanDistanceSq(p1: Point, p2: Point) i64 {
    const dx: i64 = @as(i64, p1.x) - @as(i64, p2.x);
    const dy: i64 = @as(i64, p1.y) - @as(i64, p2.y);
    const dz: i64 = @as(i64, p1.z) - @as(i64, p2.z);
    return dx * dx + dy * dy + dz * dz;
}

fn parseInput(allocator: std.mem.Allocator, filename: []const u8) ![]Point {
    const file = try std.fs.cwd().openFile(filename, .{});
    defer file.close();

    const file_size = (try file.stat()).size;
    const buffer = try allocator.alloc(u8, file_size);
    defer allocator.free(buffer);

    _ = try file.readAll(buffer);

    // Count lines
    var line_count: usize = 0;
    var line_iter1 = std.mem.splitScalar(u8, buffer, '\n');
    while (line_iter1.next()) |line| {
        if (line.len > 0) {
            line_count += 1;
        }
    }

    // Allocate points array
    const points = try allocator.alloc(Point, line_count);
    var idx: usize = 0;

    // Parse points
    var line_iter2 = std.mem.splitScalar(u8, buffer, '\n');
    while (line_iter2.next()) |line| {
        if (line.len == 0) continue;

        var it = std.mem.splitScalar(u8, line, ',');
        const x_str = it.next() orelse continue;
        const y_str = it.next() orelse continue;
        const z_str = it.next() orelse continue;

        const x = try std.fmt.parseInt(i32, x_str, 10);
        const y = try std.fmt.parseInt(i32, y_str, 10);
        const z = try std.fmt.parseInt(i32, z_str, 10);

        points[idx] = Point{ .x = x, .y = y, .z = z };
        idx += 1;
    }

    return points;
}

fn part1(allocator: std.mem.Allocator, points: []const Point, num_connections: usize) !i64 {
    const n = points.len;

    // Calculate number of pairs
    const num_pairs = (n * (n - 1)) / 2;

    // Generate all pairs with distances
    const pairs = try allocator.alloc(Pair, num_pairs);
    defer allocator.free(pairs);

    var pair_idx: usize = 0;
    for (0..n) |i| {
        for (i + 1..n) |j| {
            const dist_sq = euclideanDistanceSq(points[i], points[j]);
            pairs[pair_idx] = Pair{ .dist_sq = dist_sq, .i = i, .j = j };
            pair_idx += 1;
        }
    }

    // Sort by distance
    std.mem.sort(Pair, pairs, {}, lessThanPair);

    // Union-Find to connect closest pairs
    var uf = try UnionFind.init(allocator, n);
    defer uf.deinit();

    var connections: usize = 0;
    for (pairs) |pair| {
        _ = uf.unionSets(pair.i, pair.j);
        connections += 1;
        if (connections == num_connections) {
            break;
        }
    }

    // Get component sizes and find the 3 largest
    const sizes = try uf.getComponentSizes(allocator);
    defer allocator.free(sizes);

    // Sort sizes in descending order
    std.mem.sort(usize, sizes, {}, comptime std.sort.desc(usize));

    // Multiply the 3 largest
    const s0: i64 = @intCast(sizes[0]);
    const s1: i64 = @intCast(sizes[1]);
    const s2: i64 = @intCast(sizes[2]);
    return s0 * s1 * s2;
}

fn part2(allocator: std.mem.Allocator, points: []const Point) !i64 {
    const n = points.len;

    // Calculate number of pairs
    const num_pairs = (n * (n - 1)) / 2;

    // Generate all pairs with distances
    const pairs = try allocator.alloc(Pair, num_pairs);
    defer allocator.free(pairs);

    var pair_idx: usize = 0;
    for (0..n) |i| {
        for (i + 1..n) |j| {
            const dist_sq = euclideanDistanceSq(points[i], points[j]);
            pairs[pair_idx] = Pair{ .dist_sq = dist_sq, .i = i, .j = j };
            pair_idx += 1;
        }
    }

    // Sort by distance
    std.mem.sort(Pair, pairs, {}, lessThanPair);

    // Union-Find to connect until all in one circuit
    var uf = try UnionFind.init(allocator, n);
    defer uf.deinit();

    var num_components: usize = n;

    for (pairs) |pair| {
        if (uf.unionSets(pair.i, pair.j)) {
            num_components -= 1;
            if (num_components == 1) {
                // This was the last connection - all in one circuit now
                return @as(i64, points[pair.i].x) * @as(i64, points[pair.j].x);
            }
        }
    }

    return 0;
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const args = try std.process.argsAlloc(allocator);
    defer std.process.argsFree(allocator, args);

    const input_file = if (args.len > 1) args[1] else "../input.txt";

    const points = try parseInput(allocator, input_file);
    defer allocator.free(points);

    const part1_result = try part1(allocator, points, 1000);
    const part2_result = try part2(allocator, points);

    std.debug.print("Part 1: {d}\n", .{part1_result});
    std.debug.print("Part 2: {d}\n", .{part2_result});
}

File: 2025/day08/rust/solution.rs
use std::fs;

#[derive(Debug, Clone)]
struct Point {
    x: i32,
    y: i32,
    z: i32,
}

struct UnionFind {
    parent: Vec<usize>,
    rank: Vec<usize>,
    size: Vec<usize>,
}

impl UnionFind {
    fn new(n: usize) -> Self {
        UnionFind {
            parent: (0..n).collect(),
            rank: vec![0; n],
            size: vec![1; n],
        }
    }

    fn find(&mut self, x: usize) -> usize {
        if self.parent[x] != x {
            self.parent[x] = self.find(self.parent[x]); // Path compression
        }
        self.parent[x]
    }

    fn union(&mut self, x: usize, y: usize) -> bool {
        let mut px = self.find(x);
        let mut py = self.find(y);

        if px == py {
            return false; // Already in same set
        }

        // Union by rank
        if self.rank[px] < self.rank[py] {
            std::mem::swap(&mut px, &mut py);
        }

        self.parent[py] = px;
        self.size[px] += self.size[py];

        if self.rank[px] == self.rank[py] {
            self.rank[px] += 1;
        }

        true
    }

    fn get_component_sizes(&self) -> Vec<usize> {
        let mut sizes = Vec::new();
        for i in 0..self.parent.len() {
            if self.parent[i] == i {
                sizes.push(self.size[i]);
            }
        }
        sizes
    }
}

fn parse_input(filename: &str) -> Vec<Point> {
    let content = fs::read_to_string(filename).expect("Failed to read input file");
    content
        .lines()
        .filter_map(|line| {
            let line = line.trim();
            if line.is_empty() {
                return None;
            }
            let parts: Vec<&str> = line.split(',').collect();
            if parts.len() == 3 {
                Some(Point {
                    x: parts[0].parse().unwrap(),
                    y: parts[1].parse().unwrap(),
                    z: parts[2].parse().unwrap(),
                })
            } else {
                None
            }
        })
        .collect()
}

fn euclidean_distance_sq(p1: &Point, p2: &Point) -> i64 {
    let dx = (p1.x - p2.x) as i64;
    let dy = (p1.y - p2.y) as i64;
    let dz = (p1.z - p2.z) as i64;
    dx * dx + dy * dy + dz * dz
}

fn part1(points: &[Point], num_connections: usize) -> usize {
    let n = points.len();

    // Generate all pairs with distances
    let mut pairs: Vec<(i64, usize, usize)> = Vec::new();
    for i in 0..n {
        for j in (i + 1)..n {
            let dist_sq = euclidean_distance_sq(&points[i], &points[j]);
            pairs.push((dist_sq, i, j));
        }
    }

    // Sort by distance
    pairs.sort_by_key(|&(dist, _, _)| dist);

    // Union-Find to connect closest pairs
    let mut uf = UnionFind::new(n);
    let mut connections = 0;

    for (_, i, j) in pairs.iter() {
        uf.union(*i, *j);
        connections += 1;
        if connections == num_connections {
            break;
        }
    }

    // Get component sizes and find the 3 largest
    let mut sizes = uf.get_component_sizes();
    sizes.sort_by(|a, b| b.cmp(a)); // Sort in descending order

    // Multiply the 3 largest
    sizes[0] * sizes[1] * sizes[2]
}

fn part2(points: &[Point]) -> i64 {
    let n = points.len();

    // Generate all pairs with distances
    let mut pairs: Vec<(i64, usize, usize)> = Vec::new();
    for i in 0..n {
        for j in (i + 1)..n {
            let dist_sq = euclidean_distance_sq(&points[i], &points[j]);
            pairs.push((dist_sq, i, j));
        }
    }

    // Sort by distance
    pairs.sort_by_key(|&(dist, _, _)| dist);

    // Union-Find to connect until all in one circuit
    let mut uf = UnionFind::new(n);
    let mut num_components = n;

    for (_, i, j) in pairs.iter() {
        if uf.union(*i, *j) {
            // Actually merged two components
            num_components -= 1;
            if num_components == 1 {
                // This was the last connection - all in one circuit now
                return (points[*i].x as i64) * (points[*j].x as i64);
            }
        }
    }

    0
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let input_file = if args.len() > 1 {
        &args[1]
    } else {
        "../input.txt"
    };

    let points = parse_input(input_file);

    println!("Part 1: {}", part1(&points, 1000));
    println!("Part 2: {}", part2(&points));
}
